### PR TITLE
Add test cases to the test suite "N to N+1"

### PR DESCRIPTION
**Pull request by HoOngEe**

### CODE REVIEW

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+import * as stake from "codechain-stakeholder-sdk";
+import "mocha";
+
+import { validators as originalValidators } from "../../tendermint.dynval/constants";
+import { faucetAddress, faucetSecret } from "../helper/constants";
+import { PromiseExpect } from "../helper/promise";
+import { withNodes } from "./setup";
+import CodeChain from "../helper/spawn";
+
+async function getPossibleAuthors(
```
Comment (majecty) [inline test/src/e2e.dynval/dv.n+1.test.ts]: Other tests using stake.getPossibleAuthors, please use it.

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                termSeconds,
+                maxNumOfValidators: 8
+            }
+        });
+
+        it("betty should be included in validators", async function() {
+            const checkingNode = nodes[0];
+            const faucetSeq = await checkingNode.sdk.rpc.chain.getSeq(
+                faucetAddress
+            );
```
Comment (majecty) [inline test/src/e2e.dynval/dv.n+1.test.ts]: Let's call `beforeInsertionCheck` function here.

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+            currentTermInitialBlockNumber
+        );
+
+        expect(termMedata!.currentTermId).to.be.equals(1);
+        expect(validatorsBefore!.length).to.be.equals(
+            otherDynValidators.length
+        );
+        expect(validatorsBefore).contains.all.members(
+            otherDynValidators.map(validator => validator.platformAddress.value)
+        );
```
Comment (majecty) [inline test/src/e2e.dynval/dv.n+1.test.ts]: Let's check Betty is missing in the list.
The `contains.all` function does not guarantee that Betty is not in the list.

Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: I thought that the above length check and `contains.all` check guarantees the set identity between `validatorsBefore` and `otherDynValidators`. And `otherDynValidators` does not contain `betty`.
Is it too implicit?

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 2 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Ok, I called it.

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 1 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: I replaced it.

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+                .map((signer, index) => ({
+                    signer,
+                    delegation: 5_000,
+                    deposit: 10_000_000 - index // tie-breaker
+                }))
+                .concat([
+                    {
+                        signer: betty,
+                        delegation: 999,
+                        deposit: 10_000_000 - otherDynValidators.length
```
Comment (foriequal0) [inline test/src/e2e.dynval/dv.n+1.test.ts]: It doesn't have to subtract `otherDynValidators.length`. The tie is between nodes that have the same amount of delegation.

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+const [betty, , , ...otherDynValidators] = originalValidators;
+const allDynValidators = [betty, ...otherDynValidators];
+
+describe("Dynamic Validator N -> N+1", function() {
+    this.timeout(0);
+
+    const promiseExpect = new PromiseExpect();
+    const termSeconds = 20;
+
+    async function beforeInsertionCheck(checkingNode: CodeChain) {
```
Comment (foriequal0) [inline test/src/e2e.dynval/dv.n+1.test.ts]: `checkingNode` is only used to get `.sdk`. Would you replace it with `sdk: SDK`?

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 4 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Ah, Ok. I changed it.

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+import { validators as originalValidators } from "../../tendermint.dynval/constants";
+import { faucetAddress, faucetSecret } from "../helper/constants";
+import { PromiseExpect } from "../helper/promise";
+import { withNodes } from "./setup";
+import CodeChain from "../helper/spawn";
+
+chai.use(chaiAsPromised);
+
+const [betty, , , ...otherDynValidators] = originalValidators;
```
Comment (foriequal0) [inline test/src/e2e.dynval/dv.n+1.test.ts]: Would you change this?
```ts
const otherDynValidators = originalValidators.slice(0, 7);
const betty = originalValidators[7];
```
It is more clear that there are 7 validators in `otherDynValidators`, and it is easier than counting empty commas.

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 5 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: I changed the argument type of both `beforeInsertionCheck` and `bettyInsertionCheck`

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 3 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: I added betty check explicitly

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 6 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: I introduced slicing

**Comment by majecty** [comment]
@foriequal0 ping.

**Comment by Remagpie** [comment]
There's only one `it` for each `describe`. How about removing `describe`s here?

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 7]
```diff
@@ -146,6 +146,46 @@ describe("Dynamic Validator N -> N+1", function() {
...
+                .concat([
+                    {
+                        signer: betty,
+                        delegation: 5_000,
+                        deposit: 9999
+                    }
+                ])
+        });
+
+        it("betty shoule be included in validators", async function() {
```
Comment (Remagpie) [inline test/src/e2e.dynval/dv.n+1.test.ts]: shoule -> should

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 8]
```diff
@@ -186,6 +186,51 @@ describe("Dynamic Validator N -> N+1", function() {
...
+                        signer: betty,
+                        delegation: 999,
+                        deposit: 10_000_000
+                    }
+                ]),
+        });
+
+        it("betty should be included in validators", async function() {
+            const checkingNode = nodes[0];
+            await beforeInsertionCheck(checkingNode.sdk);
```
Comment (Remagpie) [inline test/src/e2e.dynval/dv.n+1.test.ts]: It's being called in every test suites in this file. How about moving this to `beforeEach`?

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+import { SDK } from "codechain-sdk";
+import { faucetAddress, faucetSecret } from "../helper/constants";
+import { PromiseExpect } from "../helper/promise";
+import { withNodes } from "./setup";
+
+chai.use(chaiAsPromised);
+
+const betty = originalValidators[0];
+const otherDynValidators = originalValidators.slice(1, 8);
+const allDynValidators = [betty, ...otherDynValidators];
```
Comment (Remagpie) [inline test/src/e2e.dynval/dv.n+1.test.ts]: I think it's the same as `originalValidators`.

Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Imported originalValidators has 10 members.

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 7 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Thank you I fixed

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 8 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: nodes are now being initialized in every child `describe`. To guarantee the `beforeInsertionChecking` is invoked after the nodes initialization, I should add `beforeEach` hooks for every child `describe`. Currently, I have only one `it` in every child `describe`, I thought I don't have to use `beforeEach`.
Inserting one top level `beforeEach` is not possible.

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 9 above)
Comment (Remagpie) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Then how about changing them like this?
```js
const allDynValidators = originalValidators.slice(0, 8);
const betty = allDynValidators[0];
const otherDynValidators = allDynValidators.slice(1);
```
I think it's better to explicitly state that `allDynValidators` is a subset of `originalValidators`

Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Ok, I understood it. I changed the code.

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+describe("Dynamic Validator N -> N+1", function() {
+    this.timeout(0);
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Ok, I set

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 8 above)
Comment (Remagpie) [reply test/src/e2e.dynval/dv.n+1.test.ts]: What a pity. :(

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+            validators: otherDynValidators.map((signer, index) => ({
+                signer,
+                delegation: 5_000,
+                deposit: 10_000_000 - index // tie-breaker
+            }))
+        });
+
+        it("betty should be included in validators", async function() {
+            this.slow(termSeconds * margin * 1000);
+            this.slow(termSeconds * 2 * 1000);
```
Comment (foriequal0) [inline test/src/e2e.dynval/dv.n+1.test.ts]:
```
this.slow(termSeconds * 1000);
this.timeout(termSeconds * 2 * 1000);
```

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 12]
```diff
@@ -146,6 +146,49 @@ describe("Dynamic Validator N -> N+1", function() {
         });
     });
 
+    describe("Increase one candidate's deposit which is less than the minimum deposit", async function() {
+        this.slow(termSeconds * margin * 1000);
+        this.slow(termSeconds * 2 * 1000);
```
Comment (foriequal0) [inline test/src/e2e.dynval/dv.n+1.test.ts]:
```ts
this.slow(termSeconds * 1000);
this.timeout(termSeconds * 2 * 1000);
```

File: test/src/e2e.dynval/dv.n+1.test.ts [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+            await checkingNode.waitForTx(nominateTxHash);
+
+            await checkingNode.waitForTermChange(2, termSeconds * margin);
+            await bettyInsertionCheck(checkingNode.sdk);
+        });
+    });
+
+    describe("Delegate more stake to whose stake is less than the minimum delegation", async function() {
+        this.slow(termSeconds * margin * 1000);
+        this.slow(termSeconds * 2 * 1000);
```
Comment (foriequal0) [inline test/src/e2e.dynval/dv.n+1.test.ts]:
```ts
this.slow(termSeconds * 1000);
this.timeout(termSeconds * 2 * 1000);
```

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 11 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Fixed

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 12 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Fixed

File: test/src/e2e.dynval/dv.n+1.test.ts (hunk 13 above)
Comment (HoOngEe) [reply test/src/e2e.dynval/dv.n+1.test.ts]: Fixed

**Pull request merged**